pkg/providers/vswitch/vswitch.go

```diff
@@ -0,0 +1,295 @@
+)
+
+type Provider interface {
+	LivenessProbe(*http.Request) error
+	List(context.Context, *v1alpha1.ECSNodeClass) ([]*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch, error)
+	ZonalVSwitchesForLaunch(context.Context, *v1alpha1.ECSNodeClass, []*cloudprovider.InstanceType, string) (map[string]*VSwitch, error)
+	UpdateInflightIPs(*ecs.CreateAutoProvisioningGroupRequest, *ecs.DescribeInstancesResponseBodyInstances, []*cloudprovider.InstanceType, []*VSwitch, string)
+}
+
+type DefaultProvider struct {
+	sync.Mutex
+	vpcapi                  *vpc.Client
+	cache                   *cache.Cache
+	availableIPAddressCache *cache.Cache
+	cm                      *pretty.ChangeMonitor
+	inflightIPs             map[string]int64
+}
+
+type VSwitch struct {
+	ID                      string
+	ZoneID                  string
+	AvailableIPAddressCount int64
+}
+
+func NewDefaultProvider(vpcapi *vpc.Client, cache *cache.Cache, availableIPAddressCache *cache.Cache) *DefaultProvider {
+	return &DefaultProvider{
+		vpcapi: vpcapi,
+		cm:     pretty.NewChangeMonitor(),
+		// TODO: Remove cache when we utilize the resolved vSwitches from the ECSNodeClass.status
+		// VSwitches are sorted on AvailableIpAddressCount, descending order
+		cache:                   cache,
+		availableIPAddressCache: availableIPAddressCache,
+		// inflightIPs is used to track IPs from known launched instances
+		inflightIPs: map[string]int64{},
+	}
+}
+
+func (p *DefaultProvider) List(ctx context.Context, nodeClass *v1alpha1.ECSNodeClass) ([]*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	if len(nodeClass.Spec.VSwitchSelectorTerms) == 0 {
+		return []*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch{}, nil
+	}
+	hash, err := hashstructure.Hash(nodeClass.Spec.VSwitchSelectorTerms, hashstructure.FormatV2, &hashstructure.HashOptions{SlicesAsSets: true})
+	if err != nil {
+		return nil, err
+	}
+	if switches, ok := p.cache.Get(fmt.Sprint(hash)); ok {
+		// Ensure what's returned from this function is a shallow-copy of the slice (not a deep-copy of the data itself)
+		// so that modifications to the ordering of the data don't affect the original
+		return append([]*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch{}, switches.([]*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch)...), nil
+	}
+
+	// Ensure that all the vSwitches that are returned here are unique
+	vSwitches := map[string]*vpc.DescribeVSwitchesResponseBodyVSwitchesVSwitch{}
+	for _, selectorTerms := range nodeClass.Spec.VSwitchSelectorTerms {
+		var tags []*vpc.DescribeVSwitchesRequestTag
+		var vSwitchID *string
+
+		if len(selectorTerms.ID) > 0 {
+			vSwitchID = tea.String(selectorTerms.ID)
+		}
+		for k, v := range selectorTerms.Tags {
+			// Value: nil selector all switches, '' selector specify switch
+			tag := &vpc.DescribeVSwitchesRequestTag{Key: tea.String(k)}
+			if v != "*" {
+				tag.Value = tea.String(v)
```

Comment:
I think this logic should be allowed.
If user use `abc: *`, It will be converted into a request `{key: abc, value: nil}`, this means finding all resources with the key. No matter what the value is.

Reply:
aws provider allows this?

Reply:
AWS allow. 
https://github.com/aws/karpenter-provider-aws/blob/d91930ec41695ab2667363cdc128f7a00908a253/pkg/providers/subnet/subnet.go#L267

https://docs.aws.amazon.com/de_de/AWSEC2/latest/APIReference/API_DescribeSubnets.html  Filter.N/tag-key